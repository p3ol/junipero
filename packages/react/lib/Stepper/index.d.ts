import { ReactNode, ComponentPropsWithRef } from 'react';

declare interface StepperProps extends ComponentPropsWithRef<any> {
  currentStep?: number;
  steps?: Array<object>;
  icon: ReactNode | JSX.Element;
}

declare function Stepper(props: StepperProps): ReactNode | JSX.Element;

export default Stepper;
