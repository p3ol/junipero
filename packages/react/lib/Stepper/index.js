import { Fragment } from 'react';
import { classNames } from '@junipero/react';
import PropTypes from 'prop-types';

const Stepper = ({ currentStep, steps = [], icon, ...rest }) => {
  return (
    <div {...rest}>
      <div className="junipero stepper">
        { Object.values(steps).map((step, index) => (
          <Fragment key={index}>
            <div className="step">
              <div
                className={classNames(
                  'circle',
                  index === currentStep && 'active',
                  index < currentStep && 'completed',
                  icon && 'with-icon',
                )}
              >
                { index < currentStep && icon}
              </div>
              <div className={classNames(
                'details',
                index === currentStep && 'active',
                index < currentStep && 'completed',
              )}
              >
                <span>{ step.title }</span>
                <span className="name">{ step.description }</span>
              </div>
            </div>
            <span className="divider" />
          </Fragment>
        ))}
      </div>
    </div>
  );
};

Stepper.displayName = 'Stepper';
Stepper.propTypes = {
  currentStep: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  steps: PropTypes.arrayOf(PropTypes.object),
  icon: PropTypes.node,
};

export default Stepper;
