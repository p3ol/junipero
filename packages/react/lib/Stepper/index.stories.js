import Stepper from './index';

export default { title: 'react/Stepper' };

const steps = [
  { title: 'Step 1', description: 'Description 1' },
  { title: 'Step 2', description: 'Description 2' },
  { title: 'Step 3', description: 'Description 3' },
];

export const basic = () => (
  <Stepper
    steps={steps}
    currentStep={1}
  />
);

export const withIcon = () => (
  <Stepper
    steps={steps}
    currentStep={1}
    icon={<i className="junipero-icons">check_circle</i>}
  />
);
