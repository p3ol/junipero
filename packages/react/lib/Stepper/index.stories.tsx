import Stepper from './index';
import Step from '../Step';

export default { title: 'react/Stepper' };

const steps = [
  { title: 'Step 1', content: 'Description 1' },
  { title: 'Step 2', content: 'Description 2' },
  { title: 'Step 3', content: 'Description 3' },
];

const stepsWithIcons = [
  {
    title: 'Step 1',
    content: 'Description 1',
    icon: <i className="junipero-icons">add_user</i>,
  },
  {
    title: 'Step 2',
    content: 'Description 2',
    icon: <i className="junipero-icons">user_edit</i>,
  },
  {
    title: 'Step 3',
    content: 'Description 3',
    icon: <i className="junipero-icons">user_valid</i>,
  },
];

export const Basic = () => (
  <>
    <Stepper active={0}>
      <Step title="Step 1">Description 1</Step>
      <Step title="Step 2">Description 2</Step>
      <Step title="Step 3">Description 3</Step>
    </Stepper>
    <Stepper active={1}>
      <Step title="Step 1">Description 1</Step>
      <Step title="Step 2">Description 2</Step>
      <Step title="Step 3">Description 3</Step>
    </Stepper>
    <Stepper active={2}>
      <Step title="Step 1">Description 1</Step>
      <Step title="Step 2">Description 2</Step>
      <Step title="Step 3">Description 3</Step>
    </Stepper>
    <Stepper active={3}>
      <Step title="Step 1">Description 1</Step>
      <Step title="Step 2">Description 2</Step>
      <Step title="Step 3">Description 3</Step>
    </Stepper>
  </>
);

export const WithStepsProp = () => (
  <>
    <Stepper
      steps={steps}
      active={0}
    />
    <Stepper
      steps={steps}
      active={1}
    />
    <Stepper
      steps={steps}
      active={2}
    />
    <Stepper
      steps={steps}
      active={3}
    />
  </>
);

export const WithDefaultIcon = () => (
  <>
    <Stepper
      steps={steps}
      active={0}
      icon={<i className="junipero-icons">check_circle</i>}
    />
    <Stepper
      steps={steps}
      active={1}
      icon={<i className="junipero-icons">check_circle</i>}
    />
    <Stepper
      steps={steps}
      active={2}
      icon={<i className="junipero-icons">check_circle</i>}
    />
    <Stepper
      steps={steps}
      active={3}
      icon={<i className="junipero-icons">check_circle</i>}
    />
  </>
);

export const WithCustomIcons = () => (
  <>
    <Stepper
      steps={stepsWithIcons}
      active={0}
    />
    <Stepper
      steps={stepsWithIcons}
      active={1}
    />
    <Stepper
      steps={stepsWithIcons}
      active={2}
    />
    <Stepper
      steps={stepsWithIcons}
      active={3}
    />
  </>
);
