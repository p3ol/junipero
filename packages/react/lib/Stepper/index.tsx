import {
  type RefObject,
  type ReactNode,
  Children,
  cloneElement,
  useMemo,
  useImperativeHandle,
  useRef,
  use,
} from 'react';

import type {
  JuniperoRef,
  ReactElt,
  ReactLazy,
  SpecialComponentPropsWithRef
} from '../types';
import Step, { type StepObject } from '../Step';

export declare interface StepperRef extends JuniperoRef {
  active: number;
  steps: StepObject[];
  innerRef: RefObject<HTMLDivElement>;
}

export declare interface StepperProps
  extends SpecialComponentPropsWithRef<'div', StepperRef> {
  active?: number;
  steps?: StepObject[];
  icon?: ReactNode;
}

const Stepper = ({
  ref,
  active,
  steps,
  children,
  icon,
  ...rest
}: StepperProps) => {
  const innerRef = useRef<HTMLDivElement>(null);

  useImperativeHandle(ref, () => ({
    steps,
    active,
    icon,
    isJunipero: true,
    innerRef,
  }));

  const getStepStatus = (i: number) => {
    switch (true) {
      case i < active:
        return 'completed';
      case i === active:
        return 'active';
      default:
        return null;
    }
  };

  const availableSteps = useMemo(() => (
    steps ? steps.map((t, i) => (
      <Step
        status={getStepStatus(i)}
        key={i}
        title={t.title}
        icon={t.icon || icon}
      >
        { t.content }
      </Step>
    )) : Children.toArray(children).map((
      t: ReactElt,
      i: number
    ) => (
      cloneElement(
        (t as unknown as ReactLazy ).$$typeof === Symbol.for('react.lazy')
          ? use<ReactElt>((t as unknown as ReactLazy)._payload) : t as ReactElt,
        { status: getStepStatus(i), key: i })
     ))
  ), [steps, children]);

  return (
    <div { ...rest } ref={innerRef}>
      <div className="junipero stepper">
        { availableSteps }
      </div>
    </div>
  );
};

Stepper.displayName = 'Stepper';

export default Stepper;
