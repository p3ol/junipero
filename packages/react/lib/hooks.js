import { useContext } from 'react';

import { FieldControlContext } from './contexts';

export const useFieldControl = () => useContext(FieldControlContext);
