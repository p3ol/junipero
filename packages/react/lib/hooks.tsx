import { useContext } from 'react';

import {
  AlertsContextType,
  AlertsContext,
  DropdownContext,
  DropdownContextType,
  FieldContextType,
  FieldControlContext,
  ListContext,
  ListContextType,
  ToastsContext,
  ToastsContextType,
  ModalContext,
  ModalContextType,
} from './contexts';

export const useFieldControl =
  () => useContext<FieldContextType>(FieldControlContext);
export const useDropdown = () =>
  useContext<DropdownContextType>(DropdownContext);
export const useAlerts = () => useContext<AlertsContextType>(AlertsContext);
export const useList = () => useContext<ListContextType>(ListContext);
export const useToasts = () => useContext<ToastsContextType>(ToastsContext);
export const useModal = () => useContext<ModalContextType>(ModalContext);

export { AlertsContext, DropdownContext, FieldControlContext, ListContext, ToastsContext, ModalContext };
