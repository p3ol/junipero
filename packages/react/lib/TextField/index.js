import {
  forwardRef,
  useEffect,
  useImperativeHandle,
  useReducer,
  useRef,
} from 'react';
import { classNames, mockState, exists } from '@junipero/core';
import PropTypes from 'prop-types';

import { useFieldControl } from '../hooks';

const TextField = forwardRef(({
  min,
  max,
  autoFocus,
  children,
  className,
  valid,
  value,
  disabled = false,
  required = false,
  type = 'text',
  onBlur,
  onChange,
  onFocus,
  onWheel,
  onValidate = (val, { required }) => !!val || !required,
  ...rest
}, ref) => {
  const innerRef = useRef();
  const inputRef = useRef();
  const { update: updateControl } = useFieldControl();
  const [state, dispatch] = useReducer(mockState, {
    value: value ?? '',
    valid: valid ?? false,
    focused: autoFocus ?? false,
    dirty: false,
  });

  useImperativeHandle(ref, () => ({
    innerRef,
    inputRef,
    value: state.value,
    valid: state.valid,
    dirty: state.dirty,
    focused: state.focused,
    focus,
    blur,
    reset,
    setDirty,
    isJunipero: true,
  }));

  useEffect(() => {
    if (exists(value)) {
      state.value = value;
      state.valid = onValidate(state.value, { dirty: state.dirty, required });
      dispatch({ value: state.value, valid: state.valid });
      updateControl?.({ valid: state.valid, dirty: state.dirty });
    }
  }, [value]);

  useEffect(() => {
    dispatch({ valid: valid ?? false });
    updateControl?.({ valid: state.valid });
  }, [valid]);

  const parseValue = value => {
    const value_ = value.replace(',', '.');
    const parsedValue = parseFloat(value_);

    return !isNaN(parsedValue) ? parsedValue : value_;
  };

  const onChange_ = e => {
    /* html input disabled attribute will prevent onChange if present anyway */
    /* istanbul ignore next: just in case */
    if (disabled) {
      return;
    }

    if (type === 'number') {
      value = parseValue(e?.target?.value);
    } else {
      value = e?.target?.value;
    }

    state.value = value;
    state.dirty = true;
    state.valid = onValidate(
      state.value, { dirty: state.dirty, required }
    );
    dispatch({ value: state.value, dirty: state.dirty });
    onChange?.({ value: state.value, valid: state.valid, dirty: state.dirty });
    updateControl?.({ dirty: true, valid: state.valid });
  };

  const onFocus_ = e => {
    dispatch({ focused: true });
    updateControl?.({ focused: true });
    onFocus?.(e);
  };

  const onBlur_ = e => {
    dispatch({ focused: false });
    updateControl?.({ focused: false });
    onBlur?.(e);
  };

  const focus = () => {
    inputRef.current?.focus();
  };

  const blur = () => {
    inputRef.current?.blur();
  };

  const reset = () => {
    dispatch({ value: value ?? '', valid: valid ?? false, dirty: false });
    updateControl?.({ dirty: false, valid: valid ?? false });
  };

  const isEmpty = () =>
    !exists(state.value) || state.value === '';

  const setDirty = dirty => {
    state.dirty = dirty;
    dispatch({ dirty });
    updateControl?.({ dirty });
  };

  const Tag = rest.rows > 1 ? 'textarea' : 'input';

  return (
    <div
      ref={innerRef}
      className={classNames(
        'junipero',
        'text-field',
        state.dirty ? 'dirty' : 'pristine',
        !state.valid && state.dirty ? 'invalid' : 'valid',
        { empty: isEmpty(), focused: state.focused, disabled },
        className
      )}
    >
      <Tag
        { ...rest }
        min={min}
        max={max}
        autoFocus={autoFocus}
        className="field"
        ref={inputRef}
        disabled={disabled}
        required={required}
        type={Tag === 'textarea' ? undefined : type}
        value={state.value}
        onChange={onChange_}
        onFocus={onFocus_}
        onBlur={onBlur_}
        onWheel={onWheel ?? (() => inputRef.current?.blur())}
      />
      { children }
    </div>
  );
});

TextField.displayName = 'TextField';
TextField.propTypes = {
  min: PropTypes.number,
  max: PropTypes.number,
  autoFocus: PropTypes.bool,
  disabled: PropTypes.bool,
  required: PropTypes.bool,
  type: PropTypes.string,
  valid: PropTypes.bool,
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  onBlur: PropTypes.func,
  onChange: PropTypes.func,
  onFocus: PropTypes.func,
  onWheel: PropTypes.func,
  onValidate: PropTypes.func,
};

export default TextField;
