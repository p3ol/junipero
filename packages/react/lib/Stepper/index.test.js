import { render } from '@testing-library/react';

import Stepper from './';

describe('<Stepper />', () => {

  const steps = [
    { title: 'Step 1', description: 'Description 1' },
    { title: 'Step 2', description: 'Description 2' },
    { title: 'Step 3', description: 'Description 3' },
  ];

  it('should render', () => {
    const { unmount, container } = render(
      <Stepper steps={steps} currentStep={0} />
    );
    expect(container).toMatchSnapshot();
    unmount();
  });

  it('should display an icon if the parameter is passed', () => {
    const { unmount, container } = render(
      <Stepper
        steps={steps}
        currentStep={0}
        icon={<i className="junipero-icons">check_circle</i>}
      />
    );
    expect(container).toMatchSnapshot();
    unmount();
  });
});
