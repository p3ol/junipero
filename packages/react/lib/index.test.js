import { useReducer } from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { mockState } from '@junipero/core';
import { useTimeout } from '@junipero/hooks';

import TextField from './TextField';

describe('mockState: inside react', () => {
  it('should fail update the state correctly and return the old ' +
    'state', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch({ value: { ...state.value, content: ['foo'] } });
      }, 1, []);

      const onTextChange = ({ value }) => {
        dispatch({ value: { ...state.value, text: value } });
      };

      const onContentChange = () => {
        dispatch({
          value: { ...state.value, content: [...state.value.content, 'bar'] },
        });
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content but with old text value');

    unmount();
  });

  it('should not fail update the state correctly and return the new ' +
    'state when using mockState with a callback', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch(s => ({ value: { ...s.value, content: ['foo'] } }));
      }, 1, []);

      const onTextChange = ({ value }) => {
        dispatch(s => ({ value: { ...s.value, text: value } }));
      };

      const onContentChange = () => {
        dispatch(s => ({
          value: { ...s.value, content: [...s.value.content, 'bar'] },
        }));
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content AND updated text field valud');

    unmount();
  });
});
