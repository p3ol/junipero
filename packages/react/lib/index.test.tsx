import { useReducer } from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { mockState } from '@junipero/core';
import { useTimeout } from '@junipero/hooks';

import type { FieldContent } from './types';
import TextField from './TextField';

interface State {
  value: {
    text: string;
    content: string[];
  };
}

describe('mockState: inside react', () => {
  it('should correctly update the state and return the new ' +
    'state when using mockState with a callback', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState<State>, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch(s => ({ value: { ...s.value, content: ['foo'] } }));
      }, 1, []);

      const onTextChange = ({ value }: FieldContent) => {
        dispatch(s => ({ value: { ...s.value, text: value } }));
      };

      const onContentChange = () => {
        dispatch(s => ({
          value: { ...s.value, content: [...s.value.content, 'bar'] },
        }));
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content AND updated text field valud');

    unmount();
  });
});
