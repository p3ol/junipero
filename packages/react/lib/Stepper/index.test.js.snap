// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Stepper /> should display an icon if the parameter is passed 1`] = `
<div>
  <div>
    <div
      class="junipero stepper"
    >
      <div
        class="step"
      >
        <div
          class="circle active with-icon"
        />
        <div
          class="details active"
        >
          <span>
            Step 1
          </span>
          <span
            class="name"
          >
            Description 1
          </span>
        </div>
      </div>
      <span
        class="divider"
      />
      <div
        class="step"
      >
        <div
          class="circle with-icon"
        />
        <div
          class="details"
        >
          <span>
            Step 2
          </span>
          <span
            class="name"
          >
            Description 2
          </span>
        </div>
      </div>
      <span
        class="divider"
      />
      <div
        class="step"
      >
        <div
          class="circle with-icon"
        />
        <div
          class="details"
        >
          <span>
            Step 3
          </span>
          <span
            class="name"
          >
            Description 3
          </span>
        </div>
      </div>
      <span
        class="divider"
      />
    </div>
  </div>
</div>
`;

exports[`<Stepper /> should render 1`] = `
<div>
  <div>
    <div
      class="junipero stepper"
    >
      <div
        class="step"
      >
        <div
          class="circle active"
        />
        <div
          class="details active"
        >
          <span>
            Step 1
          </span>
          <span
            class="name"
          >
            Description 1
          </span>
        </div>
      </div>
      <span
        class="divider"
      />
      <div
        class="step"
      >
        <div
          class="circle"
        />
        <div
          class="details"
        >
          <span>
            Step 2
          </span>
          <span
            class="name"
          >
            Description 2
          </span>
        </div>
      </div>
      <span
        class="divider"
      />
      <div
        class="step"
      >
        <div
          class="circle"
        />
        <div
          class="details"
        >
          <span>
            Step 3
          </span>
          <span
            class="name"
          >
            Description 3
          </span>
        </div>
      </div>
      <span
        class="divider"
      />
    </div>
  </div>
</div>
`;
